Kernel/VirtIO/VirtIOConsole.cpp

```diff
@@ -73,115 +60,135 @@ bool VirtIOConsole::handle_device_config_change()
 }
-
-bool VirtIOConsole::can_write(const FileDescription&, size_t) const
+void VirtIOConsole::process_control_message(VirtIOConsole::ControlMessage message)
 {
-    return get_queue(TRANSMITQ).has_free_slots() && m_transmit_buffer->has_space();
+    switch (message.event) {
+    case (u16)ControlEvent::DeviceAdd: {
+        auto id = message.id;
+        m_ports.set(id, make<VirtIOConsolePort>(id, *this));
+        ControlMessage ready_event {
+            .id = static_cast<u32>(id),
+            .event = (u16)ControlEvent::PortReady,
+            .value = 1 /* success */
+        };
+
+        write_control_message(ready_event);
+        break;
+    }
+    case (u16)ControlEvent::ConsolePort:
+    case (u16)ControlEvent::PortOpen: {
+        auto is_open = m_ports.get(message.id).value()->is_open();
+        if (message.value == 1 /* open */) {
+            if (!is_open) {
+                m_ports.get(message.id).value()->set_open({}, true);
+                send_open_control_message(message.id, true);
+            }
+        } else if (message.value == 0 /* close */) {
+            m_ports.get(message.id).value()->set_open({}, false);
+        } else {
+            VERIFY_NOT_REACHED();
+        }
+        break;
+    }
+    default:
+        dbgln("Unhandled message event {}!", message.event);
+        VERIFY_NOT_REACHED();
+    }
 }
-
-KResultOr<size_t> VirtIOConsole::write(FileDescription& desc, u64, const UserOrKernelBuffer& data, size_t size)
+void VirtIOConsole::write_control_message(VirtIOConsole::ControlMessage message)
 {
-    if (!size)
-        return 0;
-
-    if (!can_write(desc, size))
-        return EAGAIN;
-
-    ScopedSpinLock ringbuffer_lock(m_transmit_buffer->lock());
-    auto& queue = get_queue(TRANSMITQ);
+    ScopedSpinLock ringbuffer_lock(m_control_transmit_buffer->lock());
+    auto& queue = get_queue(CONTROL_TRANSMITQ);
     ScopedSpinLock queue_lock(queue.lock());
     VirtIOQueueChain chain(queue);
 
-    size_t total_bytes_copied = 0;
-    do {
-        PhysicalAddress start_of_chunk;
-        size_t length_of_chunk;
-
-        if (!m_transmit_buffer->copy_data_in(data, total_bytes_copied, size - total_bytes_copied, start_of_chunk, length_of_chunk)) {
-            chain.release_buffer_slots_to_queue();
-            return EINVAL;
-        }
+    PhysicalAddress start_of_chunk;
+    size_t length_of_chunk;
 
-        bool did_add_buffer = chain.add_buffer_to_chain(start_of_chunk, length_of_chunk, BufferType::DeviceReadable);
-        VERIFY(did_add_buffer);
-        total_bytes_copied += length_of_chunk;
-    } while (total_bytes_copied < size && can_write(desc, size - total_bytes_copied));
+    auto data = UserOrKernelBuffer::for_kernel_buffer((u8*)&message);
+    if (!m_control_transmit_buffer->copy_data_in(data, 0, sizeof(message), start_of_chunk, length_of_chunk)) {
+        chain.release_buffer_slots_to_queue();
+        dbgln("Failed to write data to VirtIOQueue!");
+        VERIFY_NOT_REACHED();
```

Comment:
aborting the whole system because we failed to write to a serial port doesnt sound correct.

Reply:
Yeah, the only reason this should return false is if you have too many pending items in the control queue, in which case you should just block until more space frees up.